src/cargo/util/network.rs

```diff
@@ -0,0 +1,86 @@
+use util::{CargoResult, Config, errors};
+
+/// Wrapper method for network call retry logic.
+///
+/// Retry counts provided by Config object 'net.retry'. Config shell outputs
+/// a warning on per retry.
+///
+/// Closure must return a CargoResult.
+///
+/// Example:
+/// use util::network;
+/// cargo_result = network.with_retry(&config, || something.download());
+pub fn with_retry<T, E, F>(config: &Config, mut callback: F) -> CargoResult<T>
+    where F: FnMut() -> Result<T, E>,
+          E: errors::NetworkError
+{
+    let mut remaining = try!(config.net_retry());
+    loop {
+        match callback() {
+            Ok(ret) => return Ok(ret),
+            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
+                try!(config.shell()
+                           .warn(format!("{} {} tries remaining", e,
+                                         remaining)));
```

Comment:
Perhaps this could be along the lines of:

```rust
let msg = format!("spurious network error ({} tries remaining): {}", remaining, e);
try!(config.shell().warn(msg));
```